### PR TITLE
Feature/loader

### PR DESCRIPTION
**Pull request by kjugi**
Add loader to button while clicking on it after filling up form with data. Must validate first and then showing loader and blocking button to prevent multiple clicks and sending multiple requests.

Modify button props.
Add styles to button
Create loader components with styles

### CODE REVIEW

File: view/frontend/web/js/components/BaseButton.vue [hunk 1]
```diff
@@ -1,22 +1,53 @@
 <template>
-  <button :type="buttonType" :title="buttonTitle">
-    {{ text }}
+  <button
+    class="button"
+    :type="buttonType"
+    :title="buttonTitle"
+    :disabled="isBlocked"
```
Comment (szafran89) [inline view/frontend/web/js/components/BaseButton.vue]: maybe we can set disabled button if prop withLoader is true?

File: view/frontend/web/js/components/BaseButton.vue [hunk 2]
```diff
@@ -1,22 +1,53 @@
...
-    },
-    text: {
-      type: String
-    },
-    buttonTitle: {
-      type: String
+  components: {
+    BaseLoader
+  },
+  props: ['buttonType', 'text', 'buttonTitle', 'withLoader'],
```
Comment (szafran89) [inline view/frontend/web/js/components/BaseButton.vue]: Why did you change props? Prev version was ok. We should add require property.

File: view/frontend/web/js/components/BaseButton.vue [hunk 3]
```diff
@@ -1,22 +1,53 @@
...
+    :disabled="isBlocked"
+  >
+    <div class="button__handler">
+      <span class="button__text">
+        {{ text }}
+      </span>
+
+      <template v-if="withLoader">
+        <BaseLoader/>
+      </template>
```
Comment (szafran89) [inline view/frontend/web/js/components/BaseButton.vue]: Can you move `v-if="withLoader"` to BaseLoader

File: view/frontend/web/js/components/BaseButton.vue [hunk 4]
```diff
@@ -1,22 +1,53 @@
...
+    align-items: center;
+  }
+
+  &__text {
+    display: block;
+    white-space: nowrap;
+  }
+
+  &:disabled {
+    opacity: .5;
```
Comment (szafran89) [inline view/frontend/web/js/components/BaseButton.vue]: I know it works but please change to `opacity: 0.5;`

File: view/frontend/web/js/components/BaseLoader.vue [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+      <circle cx="32" cy="4" r="4">
+        <animate
+          attributeName="opacity"
+          dur="1s"
+          values="0;1;0"
+          repeatCount="indefinite"
+          begin="0.3"
+        />
+      </circle>
+    </svg>
```
Comment (szafran89) [inline view/frontend/web/js/components/BaseLoader.vue]: please minify svg

File: view/frontend/web/js/store/index.js [hunk 6]
```diff
@@ -51,7 +51,8 @@ const store = new Vuex.Store({
       shippingCarrierCode: '',
       shippingMethodCode: ''
     },
-    totals: null
+    totals: {},
```
Comment (szafran89) [inline view/frontend/web/js/store/index.js]: why you changed to empty object?

File: view/frontend/web/js/components/BaseButton.vue (hunk 2 above)
Comment (szafran89) [reply view/frontend/web/js/components/BaseButton.vue]: Maybe it would be better to cahnge prop `withLoader` to `loader`

File: view/frontend/web/js/components/steps/TheShippingStep.vue [hunk 7]
```diff
@@ -148,9 +148,15 @@ export default {
           console.error('Error with checking email availability. \n', error)
         })
     },
+    onCountryChange (selectedOption) {
+      this.$store.commit('setLoading', true)
+      this.regions = this.$store.getters.regionsByCountryId(this.address.country_id)
+      this.$store.dispatch('updateShippingMethods', this.address.country_id)
+    },
```
Comment (szafran89) [inline view/frontend/web/js/components/steps/TheShippingStep.vue]: Do we need this method. Where do you execute it?

File: view/frontend/web/js/components/BaseButton.vue (hunk 4 above)
Comment (kjugi) [reply view/frontend/web/js/components/BaseButton.vue]: ok, added

File: view/frontend/web/js/components/BaseButton.vue (hunk 1 above)
Comment (kjugi) [reply view/frontend/web/js/components/BaseButton.vue]: prop `withLoader` is not a loader state. It's like a variant for the button. You can set button without or with loader. So I can't disable the button when `withLoader` is true b/c it will be disabled it on start

File: view/frontend/web/js/components/BaseButton.vue (hunk 2 above)
Comment (kjugi) [reply view/frontend/web/js/components/BaseButton.vue]: I think prop `loader` means it will enable loader or start the loader. In this situation, `withLoader` means it will show the button with loader option. You can also show normal button without it

File: view/frontend/web/js/components/BaseButton.vue (hunk 3 above)
Comment (kjugi) [reply view/frontend/web/js/components/BaseButton.vue]: I can say more. I removed this v-if b/c found it's unecessary

File: view/frontend/web/js/components/BaseButton.vue (hunk 2 above)
Comment (kjugi) [reply view/frontend/web/js/components/BaseButton.vue]: And I changed props b/c prev we just have `type: String` without require property

File: view/frontend/web/js/components/BaseLoader.vue (hunk 5 above)
Comment (kjugi) [reply view/frontend/web/js/components/BaseLoader.vue]: added svg-loader and add icon to assets + minify it and change styles to non scoped to get it in loader

File: view/frontend/web/js/store/index.js (hunk 6 above)
Comment (kjugi) [reply view/frontend/web/js/store/index.js]: lol idk

Comment (kjugi) [reply view/frontend/web/js/store/index.js]: And want to see it it working or no. For now we are not setting to totals anything. Is it ok?

File: view/frontend/web/js/components/BaseButton.vue (hunk 3 above)
Comment (kjugi) [reply view/frontend/web/js/components/BaseButton.vue]: nope

Comment (kjugi) [reply view/frontend/web/js/components/BaseButton.vue]: b/c it's variant for button. Not always you want to show button with loader. I.e when have 2 buttons and one is "Place order" and the second is "Back"

File: view/frontend/web/js/components/steps/TheShippingStep.vue (hunk 7 above)
Comment (kjugi) [reply view/frontend/web/js/components/steps/TheShippingStep.vue]: It looks like some dead code - deleted

File: view/frontend/web/js/components/BaseButton.vue (hunk 2 above)
Comment (szafran89) [reply view/frontend/web/js/components/BaseButton.vue]: So better is to add `require` true or false

File: view/frontend/web/js/store/index.js (hunk 6 above)
Comment (szafran89) [reply view/frontend/web/js/store/index.js]: yes, for now we don't use this state

File: view/frontend/web/js/components/steps/TheShippingStep.vue (hunk 7 above)
Comment (szafran89) [reply view/frontend/web/js/components/steps/TheShippingStep.vue]: 👍

File: view/frontend/web/js/components/BaseLoader.vue (hunk 5 above)
Comment (szafran89) [reply view/frontend/web/js/components/BaseLoader.vue]: 👍

File: view/frontend/web/js/components/BaseButton.vue (hunk 4 above)
Comment (szafran89) [reply view/frontend/web/js/components/BaseButton.vue]: 👍

File: view/frontend/web/js/components/steps/TheShippingStep.vue [hunk 8]
```diff
@@ -61,7 +61,7 @@
           In this country we don't handle any shipping methods.
         </p>
       </template>
-      <BaseButton class="button" button-type="submit" text="Next Step"/>
+      <BaseButton class="button" button-type="submit" text="Next Step" with-loader="true"/>
```
Comment (szafran89) [inline view/frontend/web/js/components/steps/TheShippingStep.vue]: you can change `with-loader="true"` to `with-loader`

File: view/frontend/web/js/components/steps/ThePaymentStep.vue [hunk 9]
```diff
@@ -50,6 +50,7 @@
       class="button"
       button-type="button"
       text="Place order"
+      with-loader="true"
```
Comment (szafran89) [inline view/frontend/web/js/components/steps/ThePaymentStep.vue]: you can change with-loader="true" to with-loader

File: view/frontend/web/js/components/steps/ThePaymentStep.vue [hunk 10]
```diff
@@ -105,13 +106,19 @@ export default {
         this.$validator.validateAll().then((result) => {
           if (result) {
             EventBus.$emit('save-address', 'billingAddress')
+            this.$store.commit('setLoading', true)
+            this.$store.commit('setAddress', {
```
Comment (szafran89) [inline view/frontend/web/js/components/steps/ThePaymentStep.vue]: Why did you add above this commit?

File: view/frontend/web/js/components/steps/TheShippingStep.vue (hunk 8 above)
Comment (kjugi) [reply view/frontend/web/js/components/steps/TheShippingStep.vue]: ok

File: view/frontend/web/js/components/steps/ThePaymentStep.vue (hunk 9 above)
Comment (kjugi) [reply view/frontend/web/js/components/steps/ThePaymentStep.vue]: ok

File: view/frontend/web/js/components/BaseButton.vue (hunk 2 above)
Comment (kjugi) [reply view/frontend/web/js/components/BaseButton.vue]: ok

**Comment by kjugi** [comment]
If we want to add loader on start than must deal with v-cloak and add some styles to loader

File: view/frontend/web/js/assets/loader.svg [hunk 11]
```diff
@@ -0,0 +1 @@
+<svg xmlns="http://www.w3.org/2000/svg"><circle cx="4" cy="4" r="4"><animate attributeName="opacity" dur="1s" values="0;1;0" repeatCount="indefinite" begin=".1"/></circle><circle cx="18" cy="4" r="4"><animate attributeName="opacity" dur="1s" values="0;1;0" repeatCount="indefinite" begin=".2"/></circle><circle cx="32" cy="4" r="4"><animate attributeName="opacity" dur="1s" values="0;1;0" repeatCount="indefinite" begin=".3"/></circle></svg>
```
Comment (szafran89) [inline view/frontend/web/js/assets/loader.svg]: missing new line :D

File: view/frontend/web/js/components/BaseButton.vue (hunk 3 above)
Comment (szafran89) [reply view/frontend/web/js/components/BaseButton.vue]: But it does not metter you can move condition to BaseLoader, like below
```
<BaseLoader  v-if="withLoader"/>
```
in this case we dont need temlate tag

**Comment by szafran89** [comment]
@kjugi just 2 comments to fix ;)

**Comment by kjugi** [comment]
Closes #26
